Add section about resource detection to Node.JS (#1767)
open-telemetry/opentelemetry.io: content/en/docs/instrumentation/js/resources.md

---
title: "Resources"
weight: 6
description:
---

> A [resource][] is an immutable representation of the entity producing telemetry
> as attributes. For example, a process producing telemetry that is running in a
> container on Kubernetes has a Pod name, it is in a namespace and possibly is
> part of a deployment which also has a name. All three of these attributes can be
> included in the resource.

Reviewer: ```suggestion
A [resource][] represents the entity producing telemetry as resource attributes.
For example, a process producing telemetry that is running in a container on
Kubernetes has a Pod name, a namespace, and possibly a deployment name.
All three of these attributes can be included in the resource.
```
Suggested copy-edits

> In your observability backend you can use the information stored in the resource
> to see to which container (or pod, or deployment, etc.) your traces belong and
> if they have issues (like out of memory, network outages, etc.) you can quickly
> understand why your traces might be slower or errounous.

Reviewer: ```suggestion
In your observability backend, you can use resource information to better investigate
interesting behavior. For example, if your trace or metrics data indicate latency in your
system, you can narrow it down to a specific container, pod, or kubernetes deployment.
```
I kind of prefer framing this up in terms of investigating behavior, but I'm pretty neutral on this suggestion.

Below you will find some introductions on how to setup resource detection with

> the Node.JS SDK

Reviewer: ```suggestion
the Node.JS SDK.
```

## Setup

Follow the instructions in the [Getting Started - Node.js][], so that you have
the files `package.json`, `app.js` and `tracing.js`.

## Process & Environment Resource Detection

> Out of the box the Node.JS SDK detects the [process and process runtime
> resources][] and takes attributes from the environment variable
> `OTEL_RESOURCE_ATTRIBUTES`. A quick way to verify that, turn on debugging in
> `tracing.js`:

Reviewer: ```suggestion
Out of the box, the Node.JS SDK detects [process and process runtime
resources][] and takes attributes from the environment variable
`OTEL_RESOURCE_ATTRIBUTES`. You can verify what it detects by
turning on diagnostic logging in `tracing.js`:
```

```javascript
// For troubleshooting, set the log level to DiagLogLevel.DEBUG
diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.DEBUG);
```

Run the application with some values set to `OTEL_RESOURCE_ATTRIBUTES`, e.g. we
set the `host.name` to identify the [Host][]:

```sh
$ env OTEL_RESOURCE_ATTRIBUTES="host.name=localhost" node --require './tracing.js' app.js
@opentelemetry/api: Registered a global for diag v1.2.0.
...
Listening for requests on http://localhost:8080
EnvDetector found resource. Resource { attributes: { 'host.name': 'localhost' } }
ProcessDetector found resource. Resource {
  attributes: {
    'process.pid': 12345,
    'process.executable.name': 'node',
    'process.command': '/app.js',
    'process.command_line': '/bin/node /app.js',
    'process.runtime.version': '16.17.0',
    'process.runtime.name': 'nodejs',
    'process.runtime.description': 'Node.js'
  }
}
...
```

> As you can see the SDK has detected the process and also added the attributes
> you set via the environment variable automatically.

Reviewer: ```suggestion
In the above example, the SDK detected the process and also added the
`host.name=localhost` attribute set via the environment variable automatically.
```

> 

Reviewer: Suggestion: add a nested subsection that makes this about adding your own resources, like:

```markdown
### Adding your own resources with environment variables
```

or

```markdown
### Adding your own resources with environment variables
```

And then that would have another section show how to do it in code instead.

Author: Great point, I missed that, updated !

Next, update your `tracing.js` like the following:

```javascript
const opentelemetry = require("@opentelemetry/sdk-node");
const {
  getNodeAutoInstrumentations,
} = require("@opentelemetry/auto-instrumentations-node");
const { diag, DiagConsoleLogger, DiagLogLevel } = require("@opentelemetry/api");
const {
  dockerCGroupV1Detector,
} = require("@opentelemetry/resource-detector-docker");

// For troubleshooting, set the log level to DiagLogLevel.DEBUG
diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.DEBUG);

const sdk = new opentelemetry.NodeSDK({
  traceExporter: new opentelemetry.tracing.ConsoleSpanExporter(),
  instrumentations: [getNodeAutoInstrumentations()],
  resourceDetectors: [dockerCGroupV1Detector],
});

sdk.start();
```

Build your docker image:

```sh
docker build . -t nodejs-otel-getting-started
```

Run your docker container:

```sh
$ docker run --rm -p 8080:8080 nodejs-otel-getting-started
@opentelemetry/api: Registered a global for diag v1.2.0.
...
Listening for requests on http://localhost:8080
DockerCGroupV1Detector found resource. Resource {
  attributes: {
    'container.id': 'fffbeaf682f32ef86916f306ff9a7f88cc58048ab78f7de464da3c320ldb5c54'
  }
}
```

> The detector has extracted the `container.id` for you. However you might
> recognize that in this example the process attributes and the attributes set via
> an environment variable are missing: By setting the `resourceDetectors` list you
> overwrite the default value. To add them back update your `tracing.js` like the
> following:

Reviewer: ```suggestion
The detector has extracted the `container.id` for you. However you might
recognize that in this example, the process attributes and the attributes set via
an environment variable are missing! To resolve this, when you set the
`resourceDetectors` list you also need to specify the `envDetector` and
`processDetector` detectors:
```

```javascript
const opentelemetry = require("@opentelemetry/sdk-node");
const {
  getNodeAutoInstrumentations,
} = require("@opentelemetry/auto-instrumentations-node");
const { diag, DiagConsoleLogger, DiagLogLevel } = require("@opentelemetry/api");
const {
  dockerCGroupV1Detector,
} = require("@opentelemetry/resource-detector-docker");
const { envDetector, processDetector } = require("@opentelemetry/resources");

// For troubleshooting, set the log level to DiagLogLevel.DEBUG
diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.DEBUG);

const sdk = new opentelemetry.NodeSDK({
  traceExporter: new opentelemetry.tracing.ConsoleSpanExporter(),
  instrumentations: [getNodeAutoInstrumentations()],

>   resourceDetectors: [envDetector, processDetector, dockerCGroupV1Detector],

Reviewer: ```suggestion
// Make sure to add all detectors you need here!
    resourceDetectors: [envDetector, processDetector, dockerCGroupV1Detector],
```

});

sdk.start();
```

Rebuild your image and run your container once again:

```shell
docker run --rm -p 8080:8080 nodejs-otel-getting-started
@opentelemetry/api: Registered a global for diag v1.2.0.
...
Listening for requests on http://localhost:8080
EnvDetector found resource. Resource { attributes: {} }
ProcessDetector found resource. Resource {
  attributes: {
    'process.pid': 1,
    'process.executable.name': 'node',
    'process.command': '/usr/src/app/app.js',
    'process.command_line': '/usr/local/bin/node /usr/src/app/app.js',
    'process.runtime.version': '18.9.0',
    'process.runtime.name': 'nodejs',
    'process.runtime.description': 'Node.js'
  }
}
DockerCGroupV1Detector found resource. Resource {
  attributes: {
    'container.id': '654d0670317b9a2d3fc70cbe021c80ea15339c4711fb8e8b3aa674143148d84e'
  }
}
...
```

> ## Visualising resources in Jaeger

Reviewer: While helpful, I'm not sure if this is something we should document here. Even though it's OSS and "blessed" by otel, jaeger is just one backend. Other backends may treat resource attributes as just fields in the schema of whatever dataset traces are sent to, for example.

Reviewer 2: I'd find it would be helpful to document the concept relations between OTEL and common backends, and how those backends are visualizing OTEL data. I've heard complaints about the confusion of misalignment of terms between OTEL and those backends. Though, this can be documented somewhere else.

Author: I removed that section.

> Though, this can be documented somewhere else.

Yes, we should have that in the Concept section or somewhere more general.
